src/main/java/io/cdap/plugin/google/sheets/sink/GoogleSheetsRecordWriter.java

```diff
@@ -0,0 +1,373 @@
+        sheetsContentShift.get(spreadSheetName).put(sheetTitle, 0);
+        sheetsRowCount.get(spreadSheetName).put(sheetTitle, sheetRowCount);
+        sheetsColumnCount.get(spreadSheetName).put(sheetTitle, sheetColumnCount);
+      }
+
+      // for each flatterned record
+      Integer sheetId = availableSheets.get(spreadSheetName).get(sheetTitle);
+      int currentShift = sheetsContentShift.get(spreadSheetName).get(sheetTitle);
+
+      // 1. prepare all needed content and merge requests
+      FlatternedRowsRequest flatternedRowsRequest =
+        sheetsSinkClient.prepareFlatternedRequest(sheetId, record, currentShift);
+      flatternedRowsRequest.setSheetTitle(sheetTitle);
+      flatternedRowsRequest.setSpreadSheetName(spreadSheetName);
+
+      // 2. update shift
+      sheetsContentShift.get(spreadSheetName).put(sheetTitle, flatternedRowsRequest.getLastRowIndex());
+
+      // 3. extend column dimension if needed
+      if (record.getHeader().getWidth() > sheetsColumnCount.get(spreadSheetName).get(sheetTitle)) {
+        int extensionSize = record.getHeader().getWidth() -
+          sheetsRowCount.get(spreadSheetName).get(sheetTitle);
+        sheetsSinkClient.extendDimension(spreadSheetId, spreadSheetName, sheetTitle, sheetId, extensionSize,
+          DimensionType.COLUMNS);
+        sheetsColumnCount.get(spreadSheetName).put(sheetTitle, record.getHeader().getWidth());
+      }
+
+      // 4. extend rows dimension if needed
+      if (sheetsContentShift.get(spreadSheetName).get(sheetTitle) >
+        sheetsRowCount.get(spreadSheetName).get(sheetTitle)) {
+        int minimumAddition = sheetsContentShift.get(spreadSheetName).get(sheetTitle) -
+          sheetsRowCount.get(spreadSheetName).get(sheetTitle);
+        int extensionSize = Math.max(minimumAddition, googleSheetsSinkConfig.getMinPageExtensionSize());
+        sheetsSinkClient.extendDimension(spreadSheetId, spreadSheetName, sheetTitle, sheetId, extensionSize,
+          DimensionType.ROWS);
+        sheetsRowCount.get(spreadSheetName).put(sheetTitle,
+          sheetsRowCount.get(spreadSheetName).get(sheetTitle) + extensionSize);
+      }
+
+      // 5. offer flatterned requests to queue
+      queueSemaphore.acquire();
+      try {
+        recordsQueue.offer(flatternedRowsRequest);
+      } finally {
+        queueSemaphore.release();
+      }
+
+      // 6. forcibly send requests to execution if the queue is exceeded
+      if (recordsQueue.size() >= googleSheetsSinkConfig.getRecordsQueueLength()) {
+        int counter = 0;
+        while (recordsQueue.size() >= googleSheetsSinkConfig.getRecordsQueueLength() && !stopSignal) {
+          if (counter > 0) {
+            Thread.sleep(1000);
+          }
+          formerScheduledExecutorService.submit(new TasksFormer(false)).get();
+          counter++;
+        }
+      }
+
+    } catch (ExecutionException | RetryException e) {
+      throw new InterruptedException(e.getMessage());
```

Comment:
this is not InterruptedException. Also add some message.

Reply:
done